chia/full_node/subscriptions.py

```diff
@@ -9,136 +9,204 @@
+
+    def peers(self, item: bytes32) -> Set[bytes32]:
+        return self._peers_for_subscription.get(item, set())
+
+    def total_count(self) -> int:
+        return len(self._peers_for_subscription)
+
+
 @dataclass(frozen=True)
 class PeerSubscriptions:
-    # TODO: use NewType all over to describe these various uses of the same types
-    # Puzzle Hash : Set[Peer ID]
-    _coin_subscriptions: Dict[bytes32, Set[bytes32]] = field(default_factory=dict, init=False)
-    # Puzzle Hash : Set[Peer ID]
-    _ph_subscriptions: Dict[bytes32, Set[bytes32]] = field(default_factory=dict, init=False)
-    # Peer ID: Set[Coin ids]
-    _peer_coin_ids: Dict[bytes32, Set[bytes32]] = field(default_factory=dict, init=False)
-    # Peer ID: Set[puzzle_hash]
-    _peer_puzzle_hash: Dict[bytes32, Set[bytes32]] = field(default_factory=dict, init=False)
-    # Peer ID: subscription count
-    _peer_sub_counter: Dict[bytes32, int] = field(default_factory=dict, init=False)
+    _ph_subscriptions: SubscriptionSet = field(default_factory=SubscriptionSet)
+    _coin_subscriptions: SubscriptionSet = field(default_factory=SubscriptionSet)
 
     def has_ph_subscription(self, ph: bytes32) -> bool:
-        return ph in self._ph_subscriptions
+        return self._ph_subscriptions.has_subscription(ph)
 
     def has_coin_subscription(self, coin_id: bytes32) -> bool:
-        return coin_id in self._coin_subscriptions
+        return self._coin_subscriptions.has_subscription(coin_id)
+
+    def peer_subscription_count(self, peer_id: bytes32) -> int:
+        ph_subscriptions = self._ph_subscriptions.count_subscriptions(peer_id)
+        coin_subscriptions = self._coin_subscriptions.count_subscriptions(peer_id)
+        return ph_subscriptions + coin_subscriptions
 
     def add_ph_subscriptions(self, peer_id: bytes32, phs: List[bytes32], max_items: int) -> Set[bytes32]:
         """
-        returns the puzzle hashes that were actually subscribed to. These may be
-        fewer than requested in case:
-        * there are duplicate puzzle_hashes
-        * some puzzle hashes are already subscribed to
-        * the max_items limit is exceeded
+        Adds subscriptions until max_items is reached. Filters out duplicates and returns all additions.
         """
 
-        puzzle_hash_peers = self._peer_puzzle_hash.setdefault(peer_id, set())
-        existing_sub_count = self._peer_sub_counter.setdefault(peer_id, 0)
-
-        ret: Set[bytes32] = set()
+        subscription_count = self.peer_subscription_count(peer_id)
+        added: Set[bytes32] = set()
 
-        # if we've reached the limit on number of subscriptions, just bail
-        if existing_sub_count >= max_items:
+        def limit_reached() -> Set[bytes32]:
             log.info(
-                "peer_id: %s reached max number of puzzle-hash subscriptions. "
-                "Not all its coin states will be reported",
+                "Peer %s reached the subscription limit.",
```

Comment:
since you have separate functions and log line, you might as well specify whether it's puzzle-hash or coin id subscriptions